spark_ada/src/mlkem.adb

```diff
@@ -1945,6 +1947,96 @@ is
    --  Exported subprogram bodies
    --=======================================
 
+   --  FIPS 203 section 7.2
+   function EK_Valid_For_Encaps (EK_Bar : in MLKEM_Encapsulation_Key)
+     return Boolean
+   is
+      Key_To_Check : constant Poly_Zq_Vector_Bytes := EK_Bar (0 .. 384 * K - 1);
+      Decoded      : NTT_Poly_Zq_Vector;
+      Reencoded    : Poly_Zq_Vector_Bytes;
+   begin
+      --  FIPS 203 7.2 - Encapsulation key check
+      --    1. Type check. Check on the length of EK is a static type-check in SPARK, so
+      --       nothing to do here.
+      --    2. Modulus check - check that Decode/Encode is idempotent:
+      Decoded := ByteDecode12 (Key_To_Check);
+      Reencoded := ByteEncode12 (Decoded);
+      return Byte_Seq_Equal (Key_To_Check, Reencoded);
+   end EK_Valid_For_Encaps;
+
+   --  FIPS 203 section 7.3 - Decapsulation key and Ciphertext check
+   function DK_Valid_For_Decaps (DK_Bar : in MLKEM_Decapsulation_Key)
+     return Boolean
+   is
+      subtype Hash_Data_Index is I32 range 0 .. 384 * K + 31;
+      subtype Hash_Data is Byte_Seq (Hash_Data_Index);
+
+      HD : constant Hash_Data := Hash_Data (DK_Bar (384 * K .. 768 * K + 31));
+      Test, Reference_Hash : Bytes_32;
+   begin
+      --  FIPS 203 7.3 - Decapsulation key check
+      --    1. Ciphertext type check. This is a static type-check in SPARK,
+      --       so nothing to do here.
+      --    2. Decapsulation key type check. This is a static type-check in
+      --       SPARK, so nothing to do here.
+      --    3. Hash check, as follows:
+      Test           := H (HD);
+      Reference_Hash := Bytes_32 (DK_Bar (768 * K + 32 .. 768 * K + 63));
+
+      return Byte_Seq_Equal (Test, Reference_Hash);
+   end DK_Valid_For_Decaps;
+
+   --  FIPS 203 section 7.1
+   function Key_Pair_Is_Consistent_With_Seed
+      (Key_Pair : in MLKEM_Key;
+       Random_D : in Bytes_32;
+       Random_Z : in Bytes_32) return Boolean
+   is
+      Regenerated_Key : MLKEM_Key;
+   begin
+      Regenerated_Key := MLKEM_KeyGen (Random_D, Random_Z);
+      return Byte_Seq_Equal (Key_Pair.EK, Regenerated_Key.EK) and
+             Byte_Seq_Equal (Key_Pair.DK, Regenerated_Key.DK);
+   end Key_Pair_Is_Consistent_With_Seed;
+
+
+   --  FIPS 203 section 7.1 "Key pair check (without seed)"
+   function Key_Pair_Check_Without_Seed
+      (Key_Pair : in MLKEM_Key;
+       Random_M : in Bytes_32) return Boolean
+   is
+      Key          : Bytes_32;
+      Key_Tick     : Bytes_32;
```

Comment:
nit: `Key` is confusing to me here, I prefer `Shared_Secret` or something like that.

Reply:
The spec (section 7.2, para 4ii - 4iii) calls these values K and K', but "K" is already used to represent the algorithm parameter K, and ' isn't legal in identifiers, so I went for Key and Key_Tick.